src/apps/virtio_net/virtio_net_dev.lua

```diff
@@ -0,0 +1,208 @@
+require('lib.virtio.virtio_h')
+
+local band, bor, rshift = bit.band, bit.bor, bit.rshift
+
+local function fstruct(def)
+   local struct = {}
+   local offset = 0
+   for ct, fld in def:gmatch('([%a_][%w_]*)%s+([%a_][%w_]*);') do
+      ct = ffi.typeof(ct)
+      struct[fld] = {
+         fieldname = fld,
+         ct = ct,
+         size = ffi.sizeof(ct),
+         offset = offset,
+      }
+      offset = offset + struct[fld].size
+   end
+   return struct, offset
+end
+
+local function fieldrd(field, fd)
+   local buf = ffi.typeof('$ [1]', field.ct)()
+   local r, err = fd:pread(buf, field.size, field.offset)
+   if not r then error(err) end
+   return buf[0]
+end
+
+local function fieldwr(field, fd, val)
+   local buf = ffi.typeof('$ [1]', field.ct)()
+   buf[0] = val
+   assert(fd:seek(field.offset))
+   local r, err = fd:write(buf, field.size)
+   if not r then error(err) end
+   return buf[0]
+end
+
+local function open_bar(fname, struct)
+   local fd, err = S.open(fname, 'rdwr')
+   if not fd then error(err) end
+   return setmetatable ({
+      fd = fd,
+      struct = struct,
+      close = function(self) return self.fd:close() end,
+   }, {
+      __index = function (self, key)
+         return fieldrd(self.struct[key], self.fd)
+      end,
+      __newindex = function (self, key, value)
+         return fieldwr(self.struct[key], self.fd, value)
+      end,
+   })
+end
+
+local virtio_pci_bar0 = fstruct[[
+   uint32_t host_features;
+   uint32_t guest_features;
+   uint32_t queue_pfn;
+   uint16_t queue_num;
+   uint16_t queue_sel;
+   uint16_t queue_notify;
+   uint8_t status;
+   uint8_t isr;
+   uint16_t config_vector;
+   uint16_t queue_vector;
+]]
+
+VGdev = {}
+VGdev.__index = VGdev
+
+function VGdev:new(args)
+   local min_features = 0 -- C.VIRTIO_F_VERSION_1?
+   local want_features = C.VIRTIO_NET_F_CSUM + C.VIRTIO_NET_F_MAC
+   pci.unbind_device_from_linux (args.pciaddr)
+
+   local bar = open_bar(pci.path(args.pciaddr..'/resource0'), virtio_pci_bar0)
+
+   bar.status = 0 -- Reset device.
+   bar.status = bor(bar.status, 1) -- Acknowledge.
+   -- Check something.
+   bar.status = bor(bar.status, 2) -- Driver.
+   local features = bar.host_features
+   if debug then print('host_features', features) end
+   if band(features, min_features) ~= min_features then
+      bar.status = bor(bar.status, 128) -- Failure.
+      bar:close()
+      return nil, "doesn't provide minimum features"
+   end
+   if debug then print('set features to:', band(features, want_features)) end
+   bar.guest_features = band(features, want_features)
+   bar.status = bor(bar.status, 8) -- Features OK.
+   if debug then print('got features: ', bar.host_features, bar.guest_features) end
+   if band(bar.status, 8) ~= 8 then
+      bar.status = bor(bar.status, 128) -- Failure.
+      bar:close()
+      return nil, "feature set wasn't accepted by device"
+   end
+
+   if debug then print("enumerating queues...") end
+   local vqs = {}
+   for qn = 0, 16 do
```

Comment:
So you try to get 16 queues here, and then go just vqs[0]/vqs[1] in the other part of the code?
Is it a multiqueue implementation or not?